perf, fix: a batch of performance improvements and fixes

milen-denev opened the pull request:
This pull request introduces several changes across the codebase, focusing on feature updates, dependency adjustments, and improvements to module loaders. The most significant updates include the addition of new hashing implementations, the replacement of the `cache` feature with `memcache`, and the addition of the `ahash` dependency. Additionally, the `get_or` method in module loaders has been updated to specify error types explicitly.

### Dependency and Feature Updates:
* Added the `ahash` dependency in `Cargo.toml` to enhance hashing capabilities.
* Replaced the `cache` feature with `memcache` in `Cargo.toml` and updated feature dependencies accordingly.
* Updated the `smallvec` dependency in `Cargo.toml` to include additional features: `union` and `const_generics`.

### Hashing Implementations:
* Added `Hash` implementations for `ServerConfigurationEntry` and `ServerConfigurationValue` to support hashing of these types. The implementation ensures deterministic hashing for `HashMap` by sorting keys before hashing. [[1]](diffhunk://#diff-3bfd9b6acac47d7c0328ff640ca433e4a8adf922f65cadda4b67795e0ce01b35R234-R252) [[2]](diffhunk://#diff-3bfd9b6acac47d7c0328ff640ca433e4a8adf922f65cadda4b67795e0ce01b35R272-R303)
* Updated `ServerConfigurationEntries` and `ServerConfigurationEntry` structs to derive `Eq` and `Hash` traits, enabling their use in hashed collections. [[1]](diffhunk://#diff-3bfd9b6acac47d7c0328ff640ca433e4a8adf922f65cadda4b67795e0ce01b35L192-R193) [[2]](diffhunk://#diff-3bfd9b6acac47d7c0328ff640ca433e4a8adf922f65cadda4b67795e0ce01b35L224-R225)

### Module Loader Improvements:
* Updated all module loader implementations to explicitly specify the error type as `anyhow::Error` in the `get_or` method for better error handling. [[1]](diffhunk://#diff-fe6dd81b4dfd55e0c7aa74957708db6412785f62aba6a44d92ca913b8de21ec2L37-R37) [[2]](diffhunk://#diff-12b461384be68483edb74465d184dd52b4a6642e34418905fb555038adabc3b1L39-R39) [[3]](diffhunk://#diff-7023055fb430aab4326af49042fe58b29812efead71a8a61d8139118accd1848L37-R37) and others)
* Removed the `cache` module and added a new `memcache` module under `optional` features in `ferron/src/modules/optional/mod.rs`. [[1]](diffhunk://#diff-793555f95ff9fc455373bbf50228a242aca020a9046ec73ba5aeedce13c63f6cL3-L4) [[2]](diffhunk://#diff-793555f95ff9fc455373bbf50228a242aca020a9046ec73ba5aeedce13c63f6cR13-R14)

### Miscellaneous Updates:
* Fixed the string slicing logic in the `get_executable` function by removing an unnecessary reference conversion.
* Added the `use std::hash::Hasher` import in `ferron/src/config/mod.rs` to support new hashing implementations.

milen-denev commented:
Please look closely at this contribution it might contain bugs. Thanks.

DorianNiemiecSVRJS commented:
I have benchmarked both upstream Ferron (what's on "ferronweb/ferron" repository) and patched Ferron (what you would like to merge into the Ferron repository), acting as a caching reverse proxy for a web application built with Node.js that responds with an "OK" message:

```
dorian@dorians-pc:~$ ferrbench -c 10000 -d 10s -t 12 -h http://localhost:8080 # Upstream Ferron
Beginning round 1...
Benchmarking 10000 connections @ http://localhost:8080 for 10 second(s)
  Latencies:
    Avg      Stdev    Min      Max      
    41.74ms  27.09ms  0.04ms   248.46ms  
  Requests:
    Total: 2380007 Req/Sec: 238870.38
  Transfer:
    Total: 428.98 MB Transfer Rate: 43.06 MB/Sec

dorian@dorians-pc:~$ ferrbench -c 10000 -d 10s -t 12 -h http://localhost:8080 # Patched Ferron
Beginning round 1...
Benchmarking 10000 connections @ http://localhost:8080 for 10 second(s)
  Latencies:
    Avg      Stdev    Min      Max      
    41.11ms  26.13ms  0.05ms   260.32ms  
  Requests:
    Total: 2403675 Req/Sec: 241854.43
  Transfer:
    Total: 433.25 MB Transfer Rate: 43.59 MB/Sec

dorian@dorians-pc:~$ ferrbench -c 10000 -d 10s -t 12 -h http://localhost:8080 # Upstream Ferron
Beginning round 1...
Benchmarking 10000 connections @ http://localhost:8080 for 10 second(s)
  Latencies:
    Avg      Stdev    Min      Max      
    41.53ms  26.57ms  0.04ms   256.08ms  
  Requests:
    Total: 2381370 Req/Sec: 239636.86
  Transfer:
    Total: 429.23 MB Transfer Rate: 43.19 MB/Sec

dorian@dorians-pc:~$ ferrbench -c 10000 -d 10s -t 12 -h http://localhost:8080 # Patched Ferron
Beginning round 1...
Benchmarking 10000 connections @ http://localhost:8080 for 10 second(s)
  Latencies:
    Avg      Stdev    Min      Max      
    41.08ms  26.80ms  0.04ms   232.18ms  
  Requests:
    Total: 2402203 Req/Sec: 241538.58
  Transfer:
    Total: 432.98 MB Transfer Rate: 43.54 MB/Sec
```

According to the benchmark results, patched Ferron has 0.8-1.2% higher throughput and 1.1-1.5% lower latency than upstream Ferron.

DorianNiemiecSVRJS commented:
The optimizations of the module cache look great!
However, I think that `should_cache_the_module` test would pass even if the "module" (which is in fact an integer value) isn't cached...
Also, all the modules use `ModuleCache`'s `get_or` method (which doesn't cache the module) instead of `get_or_init` method (which caches the module).
I have also this question - could you explain why you renamed the "cache" module to "memcache"?

milen-denev commented:
1. I don't understand the issue with should_cache_the_module. 
2. I replaced the get_or with get_or_init
3. I replaced it because I will write another module called filecache.

milen-denev commented:
> No problem, I might correct the test.
> 
> Also, I was thinking about the naming of the modules:
> 
> * the in-memory cache module would be named `cache`.
> * the file-based cache module would be named `fcache` (from "**f**ile **cache**", similarly to `rproxy` module being named after "**r**everse **proxy**").

Ok 100% agree for fcache naming.

DorianNiemiecSVRJS commented:
Thank you!
I am going to merge the pull request, and maybe correct one test, fix a Clippy warning, and rename the `memcache` module back to `cache`.
